keps/sig-apps/3541-add-enforcedRollingUpdate-to-statefulset-update-strategies/README.md

```diff
@@ -0,0 +1,679 @@
+
+#### Proposed EnforcedRollingUpdate Algorithm
+
+```
+FOR i = replicas-1 To i >= 0 DO i-- 
+    If pod[i] needs update THEN
+        wait_for_predecessors_ready(i+1 to replicas-1)
+        
+        // Key difference: proceed with update regardless of current pod state
+        If pod[i] exists THEN
+            delete_pod(i)  // Force replacement of stuck pods
+            emit_event("ForcedReplacement", pod[i])
+        ENDIF
+        create_pod(i)
+        wait_until_ready(pod[i])  // Still wait for new pod to be ready
+        
+        // Safety check: if highest ordinal pod fails, halt progression
+        if i == replicas-1 AND !pod[i].Ready THEN
+            return // Don't proceed to lower ordinals if latest fails
+        ENDIF
+    ENDIF
+ENDFOR
+```
+
+### API Changes
+
+```go
+type StatefulSetUpdateStrategyType string
+
+const (
+    RollingUpdateStatefulSetStrategyType StatefulSetUpdateStrategyType = "RollingUpdate"
+    OnDeleteStatefulSetStrategyType StatefulSetUpdateStrategyType = "OnDelete"
+    
+    // EnforcedRollingUpdateStatefulSetStrategyType indicates that update will be
+    // applied to all Pods in the StatefulSet with respect to the StatefulSet
+    // ordering constraints. When a scale operation is performed with this
+    // strategy, new Pods will be created from the specification version indicated
+    // by the StatefulSet's updateRevision. And whatever the pod status is healthy
+    // or broken, the rolling update process will not stuck.
+    EnforcedRollingUpdateStatefulSetStrategyType StatefulSetUpdateStrategyType = "EnforcedRollingUpdate"
+)
+
+// StatefulSetUpdateStrategy indicates the strategy that the StatefulSet
+// controller will use to perform updates.
+type StatefulSetUpdateStrategy struct {
+    // Type indicates the type of the StatefulSetUpdateStrategy.
+    // Default is RollingUpdate.
+    // +optional
+    Type StatefulSetUpdateStrategyType `json:"type,omitempty"`
+    // RollingUpdate is used to communicate parameters when Type is RollingUpdateStatefulSetStrategyType.
+    // +optional
+    RollingUpdate *RollingUpdateStatefulSetStrategy `json:"rollingUpdate,omitempty"`
+    // EnforcedRollingUpdate is used to communicate parameters when Type is EnforcedRollingUpdateStatefulSetStrategyType.
+    // +optional
+    EnforcedRollingUpdate *RollingUpdateStatefulSetStrategy `json:"enforcedRollingUpdate,omitempty"`
+}
+```
+
+### Implementation Changes
+
+1. Before, when updating statefulSet, we should make sure that current replicas should be Running, Ready and Available.
+Now when StatefulSetUpdateStrategyType is EnforcedRollingUpdate, we'll continue the process.
+
+2. Before, when deleting so called condemned pods who are pods beyond the updated statefulSet replicas, we'll wait for all predecessors to be Running and Ready prior to attempting a deletion.
+Now when StatefulSetUpdateStrategyType is EnforcedRollingUpdate, we'll continue the process.
```

Comment:
Is that really true? The algorithm above suggests that waiting for predecessors actually doesn't change.